Add fields.Iterable as a base class for collection field types

Fongshway opened the pull request:
Resolves https://github.com/marshmallow-code/marshmallow/issues/1263, based on what @deckar01 proposed in https://github.com/marshmallow-code/marshmallow/issues/1263#issuecomment-503755620. This PR also helps address https://github.com/marshmallow-code/marshmallow/issues/1549.

- `fields.List` is updated to inherent from `fields.Iterable`, since the majority of the `fields.List` code could be reused.
- Example of creating custom iterable field classes:

```python
from marshmallow import fields

class FrozenSet(fields.Iterable):
    iterable_type = frozenset

class ImmutableList(fields.Iterable):
    iterable_type = tuple

```

Fongshway committed:
Add fields.Iterable as a base class for collection field types

Fongshway committed:
AUTHORS and CHANGELOG

Fongshway committed:
update version in docs to match changelog

Fongshway committed:
replace f-string with format for python 3.5 compatibility

lovasoa commented:
@sloria do you have a feedback on that ? It would be useful for https://github.com/lovasoa/marshmallow_dataclass/issues/99

lafrech commented:
I just had a quick look.

Random thoughts:

- This adds a cast to each serialization/deserialization operation. Performance impact ?
- Should the type be used only on deserialization or also on serialization as in this PR ? I generally think of the field type as the type of the data in object world, not in serialized world.

Fongshway committed:
Merge remote-tracking branch 'origin/dev' into issue/1263

Fongshway committed:
only cast non list iterables

KevinMFong commented:
Thanks for the feedback @lafrech !

> This adds a cast to each serialization/deserialization operation. Performance impact ?

I updated the code to only apply the cast on each (de)serialization operation if the iterable type is not `list`. For anyone curious about the performance impact of the cast, here's what I found by doing some back of the envelop checks in IPython:
```
n = 1_000_000
%timeit foo = [i*i for i in range(n)]  # 71.8 ms ± 799 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit bar = list([i*i for i in range(n)])  # 84.2 ms ± 1.45 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit baz = list(i*i for i in range(n))  # 87.9 ms ± 2.31 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
Based on these results, I decided it was better to keep the list comprehension and only cast non-list iterables instead of using a generator expression within the cast.

> Should the type be used only on deserialization or also on serialization as in this PR ? I generally think of the field type as the type of the data in object world, not in serialized world.

I went back and forth on this, especially since there is some discussion in https://github.com/marshmallow-code/marshmallow/issues/1263 around exposing the deserialization and serialization types separately in the `Iterable` field. Here are some of my reasons for applying the type on both:
- When serializing to a Python dictionary, the least surprising behavior for users would be to use the corresponding iterable type for a given field.
- Consistency with how the `Decimal` field works, which (de)serializes to the Python `decimal.Decimal` type. And since users already need to think about which Marshmallow fields are serializable by the default `json` module, it seemed reasonable to serialize to things like sets and tuples, even if they don't have corresponding JSON types.

Fongshway commented:
Thanks for the feedback @lafrech !

> This adds a cast to each serialization/deserialization operation. Performance impact ?

I updated the code to only apply the cast on each (de)serialization operation if the iterable type is not `list`. For anyone curious about the performance impact of the cast, here's what I found by doing some back of the envelop checks in IPython:
```
n = 1_000_000
%timeit foo = [i*i for i in range(n)]  # 71.8 ms ± 799 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit bar = list([i*i for i in range(n)])  # 84.2 ms ± 1.45 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit baz = list(i*i for i in range(n))  # 87.9 ms ± 2.31 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
Based on these results, I decided it was better to keep the list comprehension and only cast non-list iterables instead of using a generator expression within the cast.

> Should the type be used only on deserialization or also on serialization as in this PR ? I generally think of the field type as the type of the data in object world, not in serialized world.

I went back and forth on this, especially since there is some discussion in https://github.com/marshmallow-code/marshmallow/issues/1263 around exposing the deserialization and serialization types separately in the `Iterable` field. Here are some of my reasons for applying the type on both:
- When serializing to a Python dictionary, the least surprising behavior for users would be to use the corresponding iterable type for a given field.
- Consistency with how the `Decimal` field works, which (de)serializes to the Python `decimal.Decimal` type. And since users already need to think about which Marshmallow fields are serializable by the default `json` module, it seemed reasonable to serialize to things like sets and tuples, even if they don't have corresponding JSON types.

Fongshway commented:
@sloria @lafrech any interest in merging this PR? If not, I'll just close it.

sloria commented:
Thanks @Fongshway and sorry for the delay in reviewing. I think this would be a good feature to have, just haven't had time to give this the attention it deserves. Appreciate your patience 🙏 

@lafrech Will you have time to review this within the next week?

Fongshway commented:
@sloria no worries and totally understand. Thank you for all of your work on this project!

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -701,7 +701,8 @@ def _serialize(
             return None
-        return [self.inner._serialize(each, attr, obj, **kwargs) for each in value]
+        result = [self.inner._serialize(each, attr, obj, **kwargs) for each in value]
+        return result if self.iterable_type == list else self.iterable_type(result)
```
Nitpick: I think the test should be 

```py
if self.iterable_type is list
```

lafrech reviewed (commented):
Could be worth adding a test showing how to subclass to use a set or a frozenset. We could even support those in the core, but this can be done in another PR.

I'm still concerned about the fact this applies to both deserialization and serialization.

As I wrote in [this comment](https://github.com/marshmallow-code/marshmallow/issues/1549#issuecomment-809123977), I generally see the field name as expressing the deserialized type, while parameters can be used to tweak serialization.

```py
class FooField(Field):  # Will deserialiaze as Foo instance.
    __init__(self,..., serialize_as="baz"):  # Will serialize as baz
        ...
```

This vision might be wrong. Other maintainers POV welcome.

Anyway, at this stage, it could be safer to use 2 different class attributes to define each type.

```py
class Iterable(Field):
    serialization_type = list
    deserialization_type = list
```

Fongshway commented:
@lafrech thanks! I'll give other maintainers a chance to opine on applying the field type for both deserialization and serialization and will plan on making updates this weekend.

Fongshway committed:
Merge branch 'dev' into issue/1263

Fongshway committed:
Update versionadded

Fongshway committed:
Use two different class attributes

Fongshway committed:
Add tests for frozenset

Fongshway committed:
Update type hints for mypy

Fongshway committed:
Update default_error_messages

Fongshway committed:
Replace class attributes with getattr

Fongshway committed:
Refactor frozenset tests

Fongshway committed:
Revert back to class attribute implementation

Fongshway committed:
Update type hints for 3.5 tests

Fongshway commented on src/marshmallow/fields.py:
```diff
@@ -710,12 +710,15 @@ def _bind_to_schema(self, field_name, schema):
+        result = [self.inner._serialize(each, attr, obj, **kwargs) for each in value]
+        if self.serialization_type is list:
+            return result
+        return self.serialization_type(result)  # type: ignore
```
I _think_ this is a `mypy` issue. Here's the error I get without the ignore:
```
src/marshmallow/fields.py:719: error: Too many arguments for "Iterable"
```

lafrech replied:
Strange. I'd be happy to get advice from someone more familiar with typing/mypy.

Fongshway committed:
fixup

Fongshway commented:
@lafrech I added tests showing how to subclass `fields.Iterable` using `frozenset` and updated the implementation to use separate `serialization_type` and `deserialization_type` class attributes. There are also a couple of other minor edits. Let me know if you have any additional feedback!

Fongshway commented:
@sloria @lafrech just wanted to follow up and see if you had any comments on my second round of edits from April.

lafrech commented on tests/frozenset_field.py:
```diff
@@ -0,0 +1,7 @@
+from marshmallow import fields
+
+
+class FrozenSet(fields.Iterable):
```
I'd put this in tests/base.py

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -730,7 +733,38 @@ def _deserialize(self, value, attr, data, **kwargs) -> typing.List[typing.Any]:
+        Does not serialize scalar values to single-item lists.
+
+    .. versionchanged:: 3.12.0
+        Inherits from Iterable instead of Field.
```
I don't think this is worth mentioning.

`Mapping` docstring mentions the version in which it was added, but `Dict` does not mention when it inherited from it.

lafrech reviewed (changes requested):
Sorry about the delay.

Thanks for your work on this.

Two minor change requests.

I don't know about the mypy issue. Help welcome.

Fongshway committed:
Merge branch 'dev' into issue/1263

Fongshway committed:
Move FrozenSet

Fongshway committed:
docs

Fongshway committed:
docs

Fongshway committed:
mypy

Fongshway commented:
> Sorry about the delay.
> 
> Thanks for your work on this.

No problem! I appreciate you taking the time to revisit this PR.

> Two minor change requests.

Done.

> I don't know about the mypy issue. Help welcome.

After sleeping on this, I think I found a better solution. Let me know what you think of the current implementation.

lafrech commented:
Thanks for the rework.

I'm not comfortable with redefining Iterable type.

```py
_T = typing.TypeVar("_T")
IterableType = typing.Union[
    typing.Type[typing.FrozenSet],
    typing.Type[typing.List],
    typing.Type[typing.Set],
    typing.Type[typing.Tuple],
]
```

Unfortunately, I can't investigate this right now.

Hint welcome, anyone.

Fongshway committed:
mypy

Fongshway committed:
add typing-extensions to setup.py requirements

Fongshway commented:
> I'm not comfortable with redefining Iterable type.

I took another stab using `Protocol`. Since `Protocol` was introduced in Python 3.8, I had to add the `typing-extensions` dependency for backwards computability. https://github.com/python/mypy/issues/5485 has some more context on why I chose to use `Protocol` instead of `Callable`.
